Buttons are disabled when there is a null in the binding chain for Command

jkoritzinsky opened the pull request:
Fixes #846.

jkoritzinsky committed:
Buttons are disabled when there is a null in the binding chain for Command.

Fixes #846.

kekekeks commented on src/Avalonia.Controls/Button.cs:
```diff
@@ -41,8 +42,9 @@ public class Button : ContentControl
         /// </summary>
-        public static readonly StyledProperty<ICommand> CommandProperty =
-            AvaloniaProperty.Register<Button, ICommand>(nameof(Command));
+        public static readonly DirectProperty<Button, ICommand> CommandProperty =
```
If I remember correctly, direct properties lack some features. So It's probably better to subscribe to property change notification here.

jkoritzinsky replied:
The subscription to the property changed notification happens in the static constructor. It still works with DirectProperties

grokys committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

grokys reviewed (changes requested):
I tried adding [a few unit tests](https://gist.github.com/grokys/1ddb942f8a0a49090eef59ec0349e452) to test this and found some problems.

The first problem is caused by [this code](https://github.com/AvaloniaUI/Avalonia/blob/e03187073879bd19d54959ca0411b2f4d322a139/src/Markup/Avalonia.Markup/Data/ExpressionObserver.cs#L242-L252) which results in the binding error not being notified on null. I think we probably need to remove this code - for the moment that will spew a load of unwanted warnings but I think we need to address that a different way in another PR.

The second problem is that if the `Command` is updated to point to a disabled `ICommand` then the button will be enabled even when it should be disabled.

jkoritzinsky committed:
Made changes requested by @grokys.

jkoritzinsky committed:
Merge branch 'DisableButtonOnNullCommandBinding' of https://github.com/jkoritzinsky/Avalonia into DisableButtonOnNullCommandBinding

grokys committed:
Merge branch 'master' into pr/894-buttons-are-disabled-when-there-is-a-null-in-the-binding-chain-for-command-3

grokys committed:
Undo bad merge of csproj.

Not sure how these changes got made.

kekekeks commented on Avalonia.sln:
```diff
@@ -1,5 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
+VisualStudioVersion = 15.0.26403.3
```
Duplicate

grokys commented:
@jkoritzinsky thanks for updating - I'm not ignoring this but I want to do a proper review and merge it after we've released alpha 5 if that's ok?

jkoritzinsky commented:
Sure!

jkoritzinsky committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

grokys committed:
Tweaked logging of binding errors.

1. Fixed some tests to expect `BindingNotification`s to be returned on a broken binding chain.
2. Changed logging of `BindingNotification`s - log at `Warning` level (instead of `Error`) except when the binding chain is broken at the root, in which case log at `Information` level. Do this to prevent flooding the output window when initializing.

grokys committed:
Fix failing test.

grokys committed:
Merge remote-tracking branch 'origin/master' into pr/894-buttons-are-disabled-when-there-is-a-null-in-the-binding-chain-for-command-3

 Conflicts:
	tests/Avalonia.Markup.UnitTests/Data/ExpressionObserverTests_Property.cs

grokys committed:
Make failing tests pass.

To do this needed to change behavior a little in that now binding errors update the target. Previously in the case of a binding error at the first node in the binding chain, we were converting the `BindingNotification` to `UnsetValue` which had the effect of updating the target value. Now we're passing the `BindingNotification` back, we need to make sure this happens. I believe this is the right thing to do as the behavior should be the same no matter where in the binding chain the error occurs. Data validation errors continue to not update the target.

grokys committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

grokys committed:
Removed duplicate line.

grokys committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

jkoritzinsky committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

jkoritzinsky committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

jkoritzinsky committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding

jkoritzinsky committed:
Merge branch 'master' into DisableButtonOnNullCommandBinding
